README
NAME
    Locale::VersionedMessages - handle all aspects of the localization
    process

SYNOPSIS
      use Locale::VersionedMessages

      $obj     = new Locale::VersionedMessages;
      $vers    = $obj->version();
      $obj->search(LOCALE_LIST);

DESCRIPTION
    Although several modules exist for handling internationalization and
    localization of a prgram, they are missing functionality for dealing
    with all of the functions to perform the task fully. Existing modules
    work fine with respect to a programmer who wants to access localized
    messages, but the modules do not contain the necessary functions to
    maintain those messages.

    This module (and the programs that are distributed with it) form a
    complete package for handling the localization problem.

    At the heart of the localization problem, you have a set of messages.
    These messages may be translated into any number of locales, and a
    programmer can then look up the appropriate message, as it would appear
    in the desired locale.

    Any number of sets of messages may be used by a program. A set of
    messages is best thought of as a set of related messages which might be
    used by any number of programs. For example, one set of messages might
    be I/O error messages encountered when working with files. Another set
    might be the messages specific to one particular program.

    Each set of messages can be translated into the language for any number
    of locales. Ideally, all messages would be translated into all locales,
    but that is often not the case. However, every set of messages has a
    default locale (which may not be the same across all sets of messages)
    which MUST have all of the messages in it.

    This allows you to look up a message in either the locale you are
    working with, or the default locale if it is not available there.

    Unlike most existing localization modules, this module keeps track of
    the version of every message in every locale. Although this information
    is not typically useful to those programmers who are simply looking up
    messages in the table, it is extremely useful to those who are
    responsible for translating and maintaing the messages.

METHODS
    The following methods are available:

    new
           $obj = new Locale::VersionedMessages;

        This creates a new Locale::VersionedMessages object.

    version
           use Locale::VersionedMessages;
           $obj = new Locale::VersionedMessages;
           $vers = $obj->version();

        Check the module version.

    err
           $err = $obj->err();

        This returns any error message set during the previous operation.

    set
           $obj->set(SET0 [,SET1,SET2,...]);

        This loads any number of message sets into the object. You can only
        look up messages from sets that have been loaded.

           @set = $obj->set();

        This returns the names of the message sets that have been loaded.
        SETi is a string containing alphanumeric and underscore characters.

    query_set_default
    query_set_locales
    query_set_msgid
        These return information about a message set.

           $locale = $obj->query_set_default();

        returns the default locale.

           @locale = $obj->query_set_locales();

        returns the list of all locales for which this message set is
        defined.

           @msgid = $obj->query_set_msgid();

        returns a list of all message IDs in this set of messages. The
        message ID is the label used in the program to access a message.

    search
        This method is used to specify the search order of the locales.

        By default, all messages are returned in the default locale of the
        message set (which may not be the same across message sets), but you
        can change this using these methods.

           $obj->search(LOCALE0 [,LOCALE1,LOCALE2,...]);

        This specifies the global search order. This is the default search
        order of locales used for all message sets. This specifies that, in
        all message sets, the default is to use LOCALE0, followed by
        LOCALE1, LOCALE2, etc.

        If the message is not found in any of the locales, the default
        locale for the set will be used.

        The global search order can be overridden on a per-set basis.

           $obj->search();

        This erases the global default search order.

           $obj->search(SET, LOCALE0 [,LOCALE1,LOCALE2,...]);

        This specifies the search order for the given set, overriding the
        global default search order. It is not necessary that all of the
        locales actually have translations. Any that don't will be silently
        ignored.

        If a message is not found in any of these locales, the default
        locale for the message set will be used.

           $obj->search(SET);

        This erases the search order for the given set, so the global search
        order will be used.

        For example:

           $obj->search('BUTTONS','en','en_US','en_GB');

        would looke for a messages in the 'BUTTONS' set in the locales:
        'en', 'en_US', and 'en_GB' in that order. If a message were not
        found in any of those three, the default locale would be used.

    query_search
           @locale = $obj->query_search();
           @locale = $obj->query_search(SET);

        This returns the global search order or the search order for the
        set.

        In the second call, the global search order is NOT returned if the
        set-specific search order is not set.

    message
           $text           = $obj(SET, MSGID [,LOCALE] [,VALUES]);
           ($text,$locale) = $obj(SET, MSGID [,LOCALE] [,VALUES]);

        This method is used to look up a message in the given set and
        substitutes in VALUES (described below in MESSAGE SUBSTITUTIONS). If
        LOCALE is present, it only looks up the message in that locale. By
        default, it will look in the search order for that set.

        If the message is not found, $text will be the empty string.

        In list context, the text will be returned along with the name of
        the locale in which the message was found.

        If any error occurs, an empty string will be returned for $text.

    query_msg_locales
    query_msg_vers
        These return information about a specific message.

           @locale = $obj->query_msg_locales(SET, MSGID);

        returns a list of all locales for which this message is defined. The
        first one will be the default locale.

           $vers = $obj->query_msg_vers(SET, MSGID [,LOCALE]);

        returns the version of the message in the given locale. If LOCALE is
        not given, it defaults to the default locale. If the message is not
        defined in the locale, a version of 0 is returned.

MESSAGE SUBSTITUTIONS
    When a message is defined, it can have values that will be passed in
    which can be inserted into the message.

    Unlike most localization tool kits, the values are passed in by name,
    rather than position in a list, so the substitution looks a bit
    different than other tool kits.

    When a message is initially defined, two pieces of information are
    required: the message ID and a list of substitution variables. Other
    information (such as a description of the message) may also be provided,
    but is not relevant to message substitution.

    If one of the substition variables is named 'foo', then anywhere in the
    message, I can include '[foo...]' and that portion of the message will
    be replaced based on the value of foo that is passed in.

    The syntax of the substitution string can be any of the following:

    [foo]
        In it's simplest form, the value is simply inserted into the string.
        For example, if the text is defined (in the default lexicon) as:

           Set:        Set1
           Message ID: Foo value [foo]
           Text:       The value of foo is [foo].

        Then:

           $obj->message('Set1','Foo value [foo]',
                         'foo' => 'bar');
              => 'The value of foo is bar.

    [foo:FORMAT]
        The value can be formatted using standard sprintf formats. Only
        simple formats are allowed, so FORMAT can be something like '%.3f'
        but not 'the number %3d'. To be exact, FORMAT must be any valid
        format that can be handled by sprintf that takes exactly one
        argument, so something like %% is not allowed, but any %d format is.

           Set:        Set1
           Message ID: Foo value [foo]
           Text:       The value of foo is >[foo:%5s]<.

           $obj->message('Set1','Foo value [foo]',
                         'foo' => 'bar');
              => 'The value of foo is >  bar<.

    [foo:quant ...] or [foo:quant:FORMAT ...]
        This is for handling plural elements. These two forms are identical
        except that the number will be formatted using a sprintf format in
        the second case.

        The general form of this is:

           [VAL:quant [COND1] [STRING1] [COND2] [STRING2] ... [DEFAULT_STRING]]

        Each CONDi is a string involving a correctly specified numerical
        test. It should be noted that the tests are NOT passed to the perl
        interpreter (via eval), so the full perl syntax is not supported.
        Since tests are specified in user (i.e. translator) defined files,
        passing portions of these files to eval would represent a security
        risk, so instead, the condition strings are manually parsed, and
        only a limited syntax is supported. However, the sytax should be
        flexible enough to handle all real-life cases.

        A condition can be a simple test. Simple tests are all of one of the
        forms:

           NUM <  NUM
           NUM <= NUM
           NUM == NUM
           NUM >= NUM
           NUM >  NUM
           NUM != NUM

        and each NUM can be:

           _VAL
           _VAL % DIGITS
           DIGITS

        Note that there are no signs allowed... only positive integers are
        allowed. Also, parentheses are not allowed inside a simple test
        (though the simple test can be enclosed in parentheses).

        Simple tests can also be combined using parentheses and the two
        operators '&&' and '||'.

        If COND1 is met, then STRING1 is used. If COND2 is met, then STRING2
        is used. If none of the conditions are met, then the DEFAULT_STRING
        is met.

        The DEFAULT_STRING is required, and at least one COND and STRING
        should be included (though they are not strictly required... if they
        are not present,t hen the DEFAULT_STRING will always be used.

        An actual example might be:

           Set:        Set1
           Message ID: Number of oranges [n]
           Text:       I have [n:quant [_n>1] [_n oranges] [_n==1] [1 orange] [no oranges]].

        and:

           $obj->message('Set1','Number of oranges [n]',1);
              => 'I have 1 orange.'

           $obj->message('Set1','Number of oranges [n]',3);
              => 'I have 3 oranges.'

           $obj->message('Set1','Number of oranges [n]',0);
              => 'I have no oranges.'

        Note that all conditions and strings are enclosed in square
        brackets, and that none of the strings may include square brackets
        in them. Also note that it is not required that '_VAL' be included
        in every one of the strings. It can be included zero, one, or even
        multiple times if desired.

    Most whitespace is ignored, so the following are quivalent:

       [foo]
       [ foo ]

    as are:

       [foo:FORMAT]
       [ foo : FORMAT ]

KNOWN PROBLEMS
    None at this point.

SEE ALSO
    Locale::VersionedMessages::Overview An overview of the
    internationalization problem with a comparison to other Locale::*
    modules

    lm_admin A command-line tool for creating and maintaing sets of messages
    and their translation tables.

LICENSE
    This script is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

AUTHOR
    Sullivan Beck (sbeck@cpan.org)



Changes
Written by:
   Sullivan Beck (sbeck@cpan.org)
Any suggestions, bug reports, or donations :-) should be sent to me.

0.90  2013-03-13
  - Initial release (beta)

0.91  2013-03-18
  - Removed some optional packages from requirement list.
  - Fixed requirement for Test::Inter so that tests would not fail.

0.92  2013-12-06
  - Removed a require line for perl 5.10 (earlier versions should work).
  - Added =encoding to POD files so UTF-8 works.
  - In the quant substitutions, the format is much nicer now (but is
      backward compatible).
  - Added Cancel buttons to the lm_gui.
  - Message descriptions are now multi-line.
  - POD files for message sets are now more complete.  POD files are
      no longer created for lexicons.
  - Documentation improvements.
  - Changed HISTORY file to new Changes specification.

0.93  2014-01-29
  - Fixed pod files to not give any errors.
  - Added copyright to the generated lexicons.
  - Cleaned up pod in generated files.

0.94  2014-01-31
  - Minor release to fix CPAN POD indexing issue (doe not affect the
    module at all).



t/lib/Locale/VersionedMessages/Sets/Test2.pm
package Locale::VersionedMessages::Sets::Test2;
####################################################
#        *** WARNING WARNING WARNING ***
#
# This file was generated, and is intended to be
# maintained automatically using the lm_admin command.
# Any changes to this file may be lost the next
# time the command is run.
####################################################

use strict;
use warnings;

our($DefaultLocale,@AllLocale,%Messages);

$DefaultLocale = 'en_US';
@AllLocale     = (qw(en_US fr_FR));

%Messages = (
   'Subst_0a' => {
      'desc'  => q{Error message (not defined in French) [aaa]},
      'vals'  => ['aaa'],
   },
   'Subst_0b' => {
      'desc'  => q{Error message (contains aaa, not bbb) [aaa]},
      'vals'  => ['aaa'],
   },
   'Subst_0c' => {
      'desc'  => q{Error message (no substitutions)},
      'vals'  => [],
   },
   'Subst_0d' => {
      'desc'  => q{Error message (must contain aaa)},
      'vals'  => ['aaa'],
   },
   'Subst_0e' => {
      'desc'  => q{Error message (invalid sprintf format)},
      'vals'  => ['aaa'],
   },
   'Subst_0f' => {
      'desc'  => q{Error message (invalid sprintf format)},
      'vals'  => ['aaa'],
   },
   'Subst_0g' => {
      'desc'  => q{Error message (no default_string)},
      'vals'  => ['n'],
   },
   'Subst_0h' => {
      'desc'  => q{Error message (invalid quant characters)},
      'vals'  => ['n'],
   },
   'Subst_0i' => {
      'desc'  => q{Error message (malformed quant test)},
      'vals'  => ['n'],
   },
   'Subst_1a' => {
      'desc'  => q{Simple substitution message #1a [aaa]},
      'vals'  => ['aaa'],
   },
   'Subst_1b' => {
      'desc'  => q{Simple substitution message #1b with duplicates [aaa]},
      'vals'  => ['aaa'],
   },
   'Subst_2a' => {
      'desc'  => q{Formatted substitution message #2a [n]},
      'vals'  => ['n'],
   },
   'Subst_3a' => {
      'desc'  => q{Quantity substitution message #3a [n]},
      'vals'  => ['n'],
   },
);

1;


t/lib/Locale/VersionedMessages/Sets/Test2/en_US.pm
package Locale::VersionedMessages::Sets::Test2::en_US;
####################################################
#        *** WARNING WARNING WARNING ***
#
# This file was generated, and is intended to be
# maintained automatically using the lm_admin command.
# Any changes to this file may be lost the next
# time the command is run.
####################################################

use strict;
use warnings;

our(%Messages);

%Messages = (
   'Subst_0a' => {
      'vers'  => 2,
      'text'  => 'Error message (not defined in French) [aaa]
',
   },
   'Subst_0b' => {
      'vers'  => 2,
      'text'  => 'Error message (contains aaa, not bbb) [aaa]
',
   },
   'Subst_0c' => {
      'vers'  => 2,
      'text'  => 'Error message (no substitutions)
',
   },
   'Subst_0d' => {
      'vers'  => 2,
      'text'  => 'Error message (should contain aaa)
',
   },
   'Subst_0e' => {
      'vers'  => 2,
      'text'  => 'Error message (format) [aaa:%]
',
   },
   'Subst_0f' => {
      'vers'  => 2,
      'text'  => 'Error message (format) [aaa:%a]
',
   },
   'Subst_0g' => {
      'vers'  => 2,
      'text'  => 'Error message (default string) [n:quant [_n==1] [one unit]]
',
   },
   'Subst_0h' => {
      'vers'  => 2,
      'text'  => 'Error message (invalid quant characters) [n:quant [_n==1 && a] [one unit] [lots]]
',
   },
   'Subst_0i' => {
      'vers'  => 2,
      'text'  => 'Error message (malformed quant test) [n:quant [_n>2 && (_n<=5))] [3-5 units] [lots]]
',
   },
   'Subst_1a' => {
      'vers'  => 2,
      'text'  => 'Substitution message 1a with value [aaa] in English.
',
   },
   'Subst_1b' => {
      'vers'  => 2,
      'text'  => 'Substitution message 1b with value [aaa] (dupl: [aaa]) in English.
',
   },
   'Subst_2a' => {
      'vers'  => 2,
      'text'  => 'Substitution message 2a with formatted value [n:%05d] in English.
',
   },
   'Subst_3a' => {
      'vers'  => 2,
      'text'  => 'Substitution message 3a with [n:quant [_n==1] [one value] [_n values]] in English.
',
   },
   'Subst_3b' => {
      'vers'  => 2,
      'text'  => q(Substitution message 3b with [n:quant _n==1 'one value' " _n values"] in English.
),
   },
);

1;


Build.PL
######################################################################
# Build.PL for Locale::VersionedMessages
######################################################################

use strict;
use warnings;
use 5.008;
use Module::Build 0.21;

my %module_prereq = (
   'perl'                     => '5.008',
   'File::Path'               => '0',
   'IO::File'                 => '0',
);

my %script_prereq = (
);

my %test_prereq = (
   'Test::Inter'              => '1.04',
   'Test::More'               => '0',
   'lib'                      => '0',
);

my %config_prereq = (
   'Module::Build'            => '0.21',
);

my %build_prereq = %config_prereq;
if ($Module::Build::VERSION < 0.4004) {
   %build_prereq = (%config_prereq,%test_prereq);
}

my $class = Module::Build->subclass(
   class  => "Module::Build::Custom",
   code   => q { sub ACTION_html { my $self = shift ; $self->SUPER::ACTION_html ( @_ ) if ( $ENV{ MAKE_HTML } ) ; } }
);

my $build = $class->new(
      module_name        => "Locale::VersionedMessages",
      dist_version       => "0.94",
      dist_abstract      => "handle all aspects of the localization process",
      license            => "perl",
      dist_author        => "Sullivan Beck <sbeck\@cpan.org>",

      requires           => \%module_prereq,
      ($Module::Build::VERSION >= 0.4004
       ? (test_requires  => \%test_prereq)
       : ()),
      ($Module::Build::VERSION >= 0.30
       ? (configure_requires  => \%config_prereq)
       : ()),
      build_requires     => \%build_prereq,
      recommends         => \%script_prereq,
      sign               => 1,
      script_files       => [
                             'bin/lm_gui',
                            ],
);

$build->create_build_script;


Makefile.PL
######################################################################
# Makefile.PL for Locale::VersionedMessages
######################################################################

use strict;
use warnings;
use 5.008;
use ExtUtils::MakeMaker 6.30;

my %module_prereq = (
   'File::Path'               => '0',
   'IO::File'                 => '0',
);

my %script_prereq = (
);

my %test_prereq = (
   'Test::Inter'              => '1.04',
   'Test::More'               => '0',
   'lib'                      => '0',
);

my %config_prereq = (
   'ExtUtils::MakeMaker'      => '6.30',
);

%module_prereq = (%module_prereq,%script_prereq);
if ($ExtUtils::MakeMaker::VERSION < 6.64) {
   %config_prereq = (%config_prereq,%test_prereq);
}

WriteMakefile(
   NAME             => "Locale::VersionedMessages",
   VERSION          => "0.94",
   ABSTRACT         => "handle all aspects of the localization process",
   LICENSE          => "perl",
   ($ExtUtils::MakeMaker::VERSION >= 6.48
    ? (MIN_PERL_VERSION => "5.008")
    : ()),
   AUTHOR           => "Sullivan Beck (sbeck\@cpan.org)",
   "dist"           => {COMPRESS=>"gzip",SUFFIX=>"gz"},
   PL_FILES         => {},
   PREREQ_PM        => \%module_prereq,
   ($ExtUtils::MakeMaker::VERSION >= 6.52
    ? (CONFIGURE_REQUIRES => \%config_prereq)
    : ()),
   ($ExtUtils::MakeMaker::VERSION >= 6.64
    ? (TEST_REQUIRES   => \%test_prereq)
    : ()),
   EXE_FILES        => [
                        'bin/lm_gui',
                       ],
);


lib/Locale/VersionedMessages.pm
package Locale::VersionedMessages;
# Copyright (c) 2010-2014 Sullivan Beck. All rights reserved.
# This program is free software; you can redistribute it and/or modify it
# under the same terms as Perl itself.

########################################################################

require 5.008;
use strict;
use warnings;

our $VERSION;
$VERSION='0.94';

########################################################################
# METHODS
########################################################################

sub new {
  my($class) = @_;

  my $self = {
              'err'    => '',
              'set'    => {},
              'mess'   => {},
              'search' => [],
             };

  bless $self, $class;

  return $self;
}

sub version {
  my($self) = @_;
  return $VERSION;
}

sub err {
   my($self) = @_;
   return $$self{'err'};
}

no strict 'refs';
sub set {
   my($self,@set) = @_;

   $$self{'err'} = '';
   if (! @set) {
      return sort keys %{ $$self{'set'} };
   }

   foreach my $set (@set) {
      if ($set !~ /^[a-zA-Z0-9_]+$/) {
         $$self{'err'} = "Set must be alphanumeric/underscore: $set";
         return;
      }

      my $m = "Locale::VersionedMessages::Sets::$set";
      eval "require $m";
      if ($@) {
         chomp($@);
         $$self{'err'} = "Unable to load set: $set: $@";
         return;
      }

      my $def_locale   = ${ "${m}::DefaultLocale" };
      my @all_locale   = @{ "${m}::AllLocale" };
      my %messages     = %{ "${m}::Messages" };

      $$self{'set'}{$set} = { 'def_loc'  => $def_locale,
                              'all_loc'  => { map {$_,1} @all_locale },
                              'messages' => \%messages,
                              'search'   => [],
                            };
   }

   return;
}
use strict 'refs';

sub query_set_default {
   my($self,$set) = @_;
   $$self{'err'} = '';

   if (! exists $$self{'set'}{$set}) {
      $$self{'err'} = "Set not loaded: $set";
      return;
   }

   return $$self{'set'}{$set}{'def_loc'};
}

sub query_set_locales {
   my($self,$set) = @_;
   $$self{'err'} = '';

   if (! exists $$self{'set'}{$set}) {
      $$self{'err'} = "Set not loaded: $set";
      return;
   }

   return sort keys %{ $$self{'set'}{$set}{'all_loc'} };
}

sub query_set_msgid {
   my($self,$set) = @_;
   $$self{'err'} = '';

   if (! exists $$self{'set'}{$set}) {
      $$self{'err'} = "Set not loaded: $set";
      return;
   }

   return sort keys %{ $$self{'set'}{$set}{'messages'} };
}

sub search {
   my($self,@locale) = @_;
   $$self{'err'} = '';

   my $set;
   if (@locale  &&  exists $$self{'set'}{$locale[0]}) {
      $set = shift(@locale);
   }

   if ($set  &&  @locale) {
      $$self{'set'}{$set}{'search'} = [@locale];

   } elsif (@locale) {
      $$self{'search'} = [@locale];

   } elsif ($set) {
      $$self{'set'}{$set}{'search'} = [];

   } else {
      $$self{'search'} = [];
   }

   return;
}

sub query_search {
   my($self,$set) = @_;
   $$self{'err'} = '';

   if ($set) {
      if (! exists $$self{'set'}{$set}) {
         $$self{'err'} = "Set not loaded: $set";
         return;
      }

      return @{ $$self{'set'}{$set}{'search'} };
   }

   return @{ $$self{'search'} };
}

no strict 'refs';
sub _load_lexicon {
   my($self,$set,$locale) = @_;
   return  if (exists $$self{'mess'}{$set}{$locale});

      if ($set !~ /^[a-zA-Z0-9_]+$/) {
         $$self{'err'} = "Set must be alphanumeric/underscore: $set";
         return;
      }

   my $m = "Locale::VersionedMessages::Sets::${set}::${locale}";
   eval "require $m";
   if ($@) {
      chomp($@);
      $$self{'err'} = "Unable to load lexicon: $set [$locale]: $@";
      return;
   }

   $$self{'mess'}{$set}{$locale} = { %{ "${m}::Messages" } };

   foreach my $msgid (sort keys %{ $$self{'mess'}{$set}{$locale} }) {
      if (! exists $$self{'set'}{$set}{'messages'}{$msgid}) {
         $$self{'err'} = "Undefined message ID in lexicon: $set [$locale $msgid]";
         return;
      }
   }
}
use strict 'refs';

sub message {
   my($self,$set,$msgid,@args) = @_;
   $$self{'err'} = '';

   # Parse arguments

   my($locale,%vals);
   if (@args  &&  @args % 2) {
      $locale = shift(@args);
      %vals   = @args;
      if (! exists $$self{'set'}{$set}{'all_loc'}{$locale}) {
         $$self{'err'} = "Set not defined in locale: $set [ $locale ]";
         return '';
      }

   } else {
      %vals  = @args;
   }

   # Look up the message.

   my @locale;
   if ($locale) {
      @locale = ($locale);

   } elsif (exists $$self{'set'}{$set}{'search'}  &&
            @{ $$self{'set'}{$set}{'search'} }) {
      @locale = @{ $$self{'set'}{$set}{'search'} };

   } elsif (exists $$self{'search'}  &&
            @{ $$self{'search'} }) {
      @locale = (@{ $$self{'search'} }, $$self{'set'}{$set}{'def_loc'});

   } else {
      @locale = ($$self{'set'}{$set}{'def_loc'});
   }

   my $message;

   foreach my $l (@locale) {
      next  if (! exists $$self{'set'}{$set}{'all_loc'}{$l});

      if (! exists $$self{'mess'}{$set}{$l}) {
         $self->_load_lexicon($set,$l);
         if ($$self{'err'}) {
            if (wantarray) {
               return ('');
            } else {
               return '';
            }
         }
      }

      if (exists $$self{'mess'}{$set}{$l}{$msgid}) {
         $locale  = $l;
         $message = $$self{'mess'}{$set}{$l}{$msgid}{'text'};
         last;
      }
   }

   if (! $message) {
      $$self{'err'} = "Message not found in specified lexicons: $msgid";
      return;
   }

   $message = $self->_fix_message($set,$msgid,$message,$locale,%vals);

   if (wantarray) {
      return ($message,$locale);
   } else {
      return $message;
   }
}

sub query_msg_locales {
   my($self,$set,$msgid) = @_;
   $$self{'err'} = '';

   if (! exists $$self{'set'}{$set}) {
      $$self{'err'} = "Set not loaded: $set";
      return ();
   }
   if (! exists $$self{'set'}{$set}{'messages'}{$msgid}) {
      $$self{'err'} = "Message ID not defined in set: $set [$msgid]";
      return ();
   }

   my %all_loc = %{ $$self{'set'}{$set}{'all_loc'} };
   my $def_loc = $$self{'set'}{$set}{'def_loc'};
   delete $all_loc{$def_loc};

   my @locale = ($def_loc);
   foreach my $locale (sort keys %all_loc) {
      $self->_load_lexicon($set,$locale);
      return ()  if ($$self{'err'});
      if (exists $$self{'mess'}{$set}{$locale}{$msgid}) {
         push(@locale,$locale);
      }
   }

   return @locale;
}

sub query_msg_vers {
   my($self,$set,$msgid,$locale) = @_;
   $$self{'err'} = '';

   if (! exists $$self{'set'}{$set}) {
      $$self{'err'} = "Set not loaded: $set";
      return '';
   }
   if (! exists $$self{'set'}{$set}{'messages'}{$msgid}) {
      $$self{'err'} = "Message ID not defined in set: $set [$msgid]";
      return '';
   }

   $locale = $$self{'set'}{$set}{'def_loc'}  if (! $locale);

   if (! exists $$self{'set'}{$set}{'all_loc'}{$locale}) {
      $$self{'err'} = "Lexicon not available for set: $set [$locale]";
      return '';
   }

   if (exists $$self{'mess'}{$set}{$locale}{$msgid}) {
      return $$self{'mess'}{$set}{$locale}{$msgid}{'vers'};
   }
   return 0;
}

########################################################################
# MESSAGE SUBSTITUTIONS
########################################################################

# This takes a message and performs substitutions for each of
# the different substitution values.
#
sub _fix_message {
   my($self,$set,$msgid,$message,$locale,%vals) = @_;

   # No substitutions.

   my @vals;
   if (exists $$self{'set'}{$set}{'messages'}{$msgid}{'vals'}) {
      @vals = @{ $$self{'set'}{$set}{'messages'}{$msgid}{'vals'} };
   }
   if (! @vals) {
      if (%vals) {
         $$self{'err'} = "Message does not contain substitutions, but " .
                         "values were supplied: $msgid";
         return '';
      }
      return $message;
   }

   # Check each substitution.

   foreach my $val (sort @vals) {
      my $done;
      if (! exists $vals{$val}) {
         $$self{'err'} = "A required substitution value was not passed in: " .
                         "$msgid [$val]";
         return '';
      }
      ($message,$done) = $self->_substitute($set,$msgid,$locale,
                                            $message,$val,$vals{$val});
      return ''  if ($$self{'err'});
      if (! $done) {
         $$self{'err'} = "The message in a lexicon does not contain a required " .
                         "substitution: $msgid [$locale $val]";
         return '';
      }
      delete $vals{$val};
   }
   foreach my $val (sort keys %vals) {
      $$self{'err'} = "An invalid value was passed in: $msgid [$val]";
      return '';
   }

   return $message;
}

# This does the acutal substitution for a single substitution value.
#
# If the substitution is found, $done = 1 will be returned.
#
sub _substitute {
   my($self,$set,$msgid,$locale,$message,$var,$val) = @_;
   my $done     = 0;

   # Simple substitutions: [foo]

   if ($message =~ s/\[\s*$var\s*\]/$val/sg) {
      $done = 1;
   }

   # Formatted substitutions: [ foo : FORMAT ]

   my $fmt_re = qr/\s*:\s*(%.*?)/;

   while ($message =~ s/\[\s*$var$fmt_re\s*\]/__L_M_TMP__/s) {
      my $fmt = $1;

      no warnings;
      $val = sprintf($fmt,$val);
      use warnings;
      if ($val eq $fmt) {
         $$self{'err'} = "Invalid sprintf format: $msgid [$locale $var]";
         return;
      }
      $message =~ s/__L_M_TMP__/$val/s;
      $done    = 1;
   }

   # Quant substitutions: [ foo : quant [ : FORMAT ] ... ]

   my ($msg,$d) = $self->_quant($set,$msgid,$locale,$message,$var,$val);
   return       if ($$self{'err'});
   $message     = $msg;
   $done        = $d  if ($d);

   return ($message,$done);
}

# This tests a string for any quant substitutions and returns the
# string.
#
sub _quant {
   my($self,$set,$msgid,$locale,$mess,$var,$val) = @_;

   my $val_orig = $val;
   my $fmt_re   = qr/\s*:\s*(%.*?)/;
   my $brack_re = qr/\s+\[([^]]*?)\]/;
   my $sq_re    = qr/\s+'([^']*?)'/;
   my $dq_re    = qr/\s+"([^"]*?)"/;
   my $ws_re    = qr/\s+(\S*)/;
   my $tok_re   = qr/(?:$brack_re|$sq_re|$dq_re|$ws_re)/;
   my $tmp      = '__L_M_TMP__';
   my $done     = 0;

   SUBST:
   while ($mess =~ s/\[\s*$var\s*:\s*quant\s*(?:$fmt_re)?($tok_re+)\s*\]/$tmp/s) {
      my $fmt   = $1;
      my $tokens= $2;

      if ($fmt) {
         no warnings;
         $val = sprintf($fmt,$val);
         use warnings;
         if ($val eq $fmt) {
            $$self{'err'} = "Invalid sprintf format: $msgid [$locale $var]";
            return;
         }
      }

      my @tok;
      while ($tokens =~ s/^$tok_re//) {
         push(@tok, $1 || $2 || $3 || $4);
      }

      if (@tok % 2 == 0) {
         $$self{'err'} = "Default string required in quant substitution: " .
                         "$msgid [$locale $var]";
         return;
      }

      # @tok is (TEST, STRING, TEST, STRING, ..., DEFAULT_STRING)

      while (@tok) {
         my $ele = shift(@tok);

         # DEFAULT_STRING

         if (! @tok) {
            $ele  =~ s/_$var/$val/g;
            $mess =~ s/$tmp/$ele/s;
            $done = 1;
            next SUBST;
         }

         # TEST, STRING

         my $flag = $self->_test($set,$msgid,$locale,$var,$val_orig,$ele);
         return  if ($$self{'err'});

         $ele     = shift(@tok);
         if ($flag) {
            $ele  =~ s/_$var/$val/g;
            $mess =~ s/$tmp/$ele/s;
            $done = 1;
            next SUBST;
         }
      }
   }

   return($mess,$done);
}

# This parses a condition string and returns 1 if the condition is true for
# this value.
#
sub _test {
   my($self,$set,$msgid,$locale,$var,$n,$test) = @_;

   # $n must be an unsigned integer

   if ($n !~ /^\d+$/) {
      $$self{'err'} = "Quantity test requires an unsigned integer: " .
                      "$msgid [$locale $var]";
      return;
   }

   # Currently, test can only have:
   #   whitespace
   #   _VAR
   #   ( ) && || < <= == != >= >
   #   DIGITS
   # in them.

   my $tmp = $test;
   $tmp    =~ s/(?:\s|\d|_$var|[()&|<=!>])//g;
   if ($tmp) {
      $$self{'err'} = "Quantity test contains invalid characters: " .
                      "$msgid [$locale $var]";
      return;
   }

   # Parse the tests.

   #  1) _VAR => $n

   $test =~ s/\s*_$var\s*/ $n /g;

   #  2) DIGITS % DIGITS => DIGITS

   while ($test =~ s/\s*(\d+)\s*%\s*(\d+)\s*/__L_M_TMP__/) {
      my $m = $1 % $2;
      $test =~ s/__L_M_TMP__/ $m /;
   }

   #  3) DIGITS OP DIGITS => 0 or 1

   while ($test =~ s/\s*(\d+)\s*(==|>=|<=|!=|>|<)\s*(\d+)\s*/__L_M_TMP__/) {
      my($m,$op,$n) = ($1,$2,$3);
      my $x;
      if      ($op eq '==') {
         $x = ($m==$n ? 1 : 0);
      } elsif ($op eq '!=') {
         $x = ($m!=$n ? 1 : 0);
      } elsif ($op eq '>=') {
         $x = ($m>=$n ? 1 : 0);
      } elsif ($op eq '<=') {
         $x = ($m<=$n ? 1 : 0);
      } elsif ($op eq '>') {
         $x = ($m>$n ? 1 : 0);
      } elsif ($op eq '<') {
         $x = ($m<$n ? 1 : 0);
      }
      $test =~ s/__L_M_TMP__/$x/;
   }

   #  Repeat until done:
   #  4) (DIGITS) => DIGITS
   #  5) DIGITS BOP DIGITS => 0 or 1

   while (1) {
      my $done = 1;
      if ($test =~ s/\s*\(\s*(\d+)\s*\)\s*/$1/g) {
         $done = 0;
      }
      while ($test =~ s/\s*(\d+)\s*(\|\||&&)\s*(\d+)\s*/__L_M_TMP__/) {
         my($m,$op,$n) = ($1,$2,$3);
         my $x;
         if      ($op eq '&&') {
            $x = ($m && $n ? 1 : 0);
         } elsif ($op eq '||') {
            $x = ($m || $n ? 1 : 0);
         }
         $test =~ s/__L_M_TMP__/$x/;
         $done = 0;
      }
      last  if ($done);
   }

   # Final check:
   #   6) DIGITS => 0 or 1

   if ($test =~ /^\s*(\d+)\s*$/) {
      return ($1 ? 1 : 0);
   }

   $$self{'err'} = "Quantity test malformed: $msgid [$locale $var]";
   return;
}

1;
# Local Variables:
# mode: cperl
# indent-tabs-mode: nil
# cperl-indent-level: 3
# cperl-continued-statement-offset: 2
# cperl-continued-brace-offset: 0
# cperl-brace-offset: 0
# cperl-brace-imaginary-offset: 0
# cperl-label-offset: 0
# End:
